### ISSUE TITLE
Celery 3.1.19 MongoDB backend using wrong type for "background" argument of ensure_index call

### ISSUE THREAD

**Issue by pandasuit** [celery/celery#3011]
Celery 3.1.19
pymongo 2.8
MongoDB 2.6.10

Line in question:
https://github.com/celery/celery/blob/3.1/celery/backends/mongodb.py#L250

The MongoDB backend adds an index on the date_done field using the pymongo ensure_index method but passes a string value to the background parameter rather than the expected boolean. It should be True instead of 'true'.

I didn't notice the issue originally since pymongo and MongoDB seem to allow the string value and Celery works regardless however tools like Robomongo could not show the index and instead logged an error about a type mismatch.

The documentation of pymongo is not really that clear about types and regarding the background argument only says "if True this index should be created in the background."

I noticed that in the master branch there is another method (group_collection) that creates this index and also uses 'true' instead of True.

**Comment by ask** [comment]
So it should be set to `"True"`, instead of `"true"`?

Is that change backwards compatible?

**Comment by pandasuit** [comment]
As far as I can tell MongoDB has always expected a boolean value for the index background option. MongoDB appears to have added background indexes in 1.4 but the oldest documentation on their site is for 2.2 and in that documentation they indicate it should be a boolean. I am not sure whether earlier versions of MongoDB allowed or recommended a string. Docs:
https://docs.mongodb.org/v2.2/core/indexes/#index-creation-background

Pymongo did not explicitly mention this option until version 1.11 but does not specify a type in the early documentation and the only description is "if this index should be created in the background". Pymongo 2.7.1 was the first version to mention that the boolean True is the value that indicates the index should be created in the background but even that version doesn't prevent you from supplying a string argument. The most recent documentation doesn't make it any clearer. Docs:
https://api.mongodb.org/python/1.11/api/pymongo/collection.html#pymongo.collection.Collection.ensure_index
https://api.mongodb.org/python/2.7.1/api/pymongo/collection.html#pymongo.collection.Collection.ensure_index
Current docs only show option in the create_index method:
https://api.mongodb.org/python/current/api/pymongo/collection.html#pymongo.collection.Collection.create_index

So long story short I cannot say with 100% accuracy whether it is backwards compatible but there is no indication I can find that the string 'true' was ever a valid value and the Python boolean value True works in the versions I indicated in my original comment. Likely only testing of the various versions would prove that it is backwards compatible.

In our project I have subclassed the Celery MongoDB backend to change the way saving is done and I ended up overriding the collection() method to fix the index creation in our production code so there is a path for those of us that run into this issue to fix it on our end.

**Comment by ask** [comment]
Closing this, as we don't have the resources to support MongoDB.
